java/azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/PartitionReceiver.java

```diff
@@ -312,41 +306,69 @@ public void setReceiveHandler(final PartitionReceiveHandler receiveHandler)
 	 * @param receiveHandler An implementation of {@link PartitionReceiveHandler}
 	 * @param invokeWhenNoEvents flag to indicate whether the {@link PartitionReceiveHandler#onReceive(Iterable)} should be invoked when the receive call times out
 	 */
-	public void setReceiveHandler(final PartitionReceiveHandler receiveHandler, final boolean invokeWhenNoEvents)
+	public CompletableFuture<Void> setReceiveHandler(final PartitionReceiveHandler receiveHandler, final boolean invokeWhenNoEvents)
 	{
-		synchronized (this.receiveHandlerSync)
+		synchronized (this.receiveHandlerLock)
 		{
-			this.invokeHandlerOnTimeout = invokeWhenNoEvents;
-			
+			// user setting receiveHandler==null should stop the pump if its running
 			if (receiveHandler == null)
 			{
-				if (this.onReceiveHandler != null)
+				if (this.receivePump != null && this.receivePump.isRunning())
 				{
-					this.isOnReceivePumpRunning = false;
-					this.onReceivePumpThread.interrupt();
+					return this.receivePump.stop();
 				}
 			}
 			else
 			{
-				if (this.isOnReceivePumpRunning)
+				if (this.receivePump != null && this.receivePump.isRunning())
 					throw new IllegalArgumentException(
 					"Unexpected value for parameter 'receiveHandler'. PartitionReceiver was already registered with a PartitionReceiveHandler instance. Only 1 instance can be registered.");
 
-				this.onReceiveHandler = receiveHandler;
-				this.startOnReceivePump();
+				this.receivePump = new ReceivePump(
+					new ReceivePump.IPartitionReceiver()
+					{
+						@Override
+						public Iterable<EventData> receive(int maxBatchSize) throws ServiceBusException
+						{
+							return PartitionReceiver.this.receiveSync(maxBatchSize);
+						}
+						
+						@Override
+						public String getPartitionId()
+						{
+							return PartitionReceiver.this.getPartitionId();
+						}
+					},
+					receiveHandler,
+					invokeWhenNoEvents);
+
+				final Thread onReceivePumpThread = new Thread(new Runnable()
+					{
+						@Override
+						public void run()
+						{
+							receivePump.run();
+						}
+					});
+				
+				onReceivePumpThread.start();
 			}
+
+			return CompletableFuture.completedFuture(null);
 		}
 	}
 
 	@Override
 	public CompletableFuture<Void> onClose()
 	{
-		this.isOnReceivePumpRunning = false;
-		if (this.onReceivePumpThread != null && !this.onReceivePumpThread.isInterrupted())
+		if (this.receivePump != null && this.receivePump.isRunning())
 		{
-			this.onReceivePumpThread.interrupt();
```

Comment:
>this.onReceivePumpThread.interrupt(); [](start = 3, length = 37)

got rid of this in the new implementation - as this will not fit in the future when we move to - user ThreadPool